provider/adapters/src/fallback/adapter.rs

```diff
@@ -3,25 +3,186 @@
+    }
 }
 
-pub struct LocaleFallbackAdapter<P>(pub P);
+impl<P> LocaleFallbackProvider<P> {
+    /// Wrap a provider with an arbitrary fallback engine.
+    ///
+    /// This relaxes the requirement that the wrapped provider contains its own fallback data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu_provider::hello_world::*;
+    /// use icu_provider::prelude::*;
+    /// use icu_provider_adapters::fallback::{LocaleFallbacker, LocaleFallbackProvider};
+    ///
+    /// let provider = HelloWorldProvider::new_with_placeholder_data();
+    ///
+    /// let req = DataRequest {
+    ///     options: locale!("de-CH").into(),
+    ///     metadata: Default::default(),
+    /// };
+    ///
+    /// // There is no "de-CH" data in the `HelloWorldProvider`
+    /// ResourceProvider::<HelloWorldV1Marker>::load_resource(&provider, &req)
+    ///     .expect_err("No data for de-CH");
+    ///
+    /// // `HelloWorldProvider` does not contain fallback data,
+    /// // but we can fetch it from `icu_testdata`, and then
+    /// // use it to create the fallbacking data provider.
+    /// let fallbacker = LocaleFallbacker::try_new(&icu_testdata::get_provider())
+    ///     .expect("Fallback data present");
+    /// let provider = LocaleFallbackProvider::new_with_fallbacker(provider, fallbacker);
+    ///
+    /// // Now we can load the "de-CH" data via fallback to "de".
+    /// let german_hello_world: DataPayload<HelloWorldV1Marker> = provider
+    ///     .load_resource(&req)
+    ///     .expect("Loading should succeed")
+    ///     .take_payload()
+    ///     .expect("Data should be present");
+    ///
+    /// assert_eq!("Hallo Welt", german_hello_world.get().message);
+    /// ```
+    pub fn new_with_fallbacker(provider: P, fallbacker: LocaleFallbacker) -> Self {
+        Self {
+            inner: provider,
+            fallbacker,
+        }
+    }
+}
+
+impl<P> AnyProvider for LocaleFallbackProvider<P>
+where
+    P: AnyProvider,
+{
+    fn load_any(&self, key: ResourceKey, base_req: &DataRequest) -> Result<AnyResponse, DataError> {
+        let key_fallbacker = self.fallbacker.for_key(key);
+        let mut fallback_iterator = key_fallbacker.fallback_for(base_req.clone());
+        while !fallback_iterator.get().options.is_empty() {
+            let result = self.inner.load_any(key, fallback_iterator.get());
+            if !result_is_err_missing_resource_options(&result) {
+                return result;
+            }
+            fallback_iterator.step();
+        }
+        Err(DataErrorKind::MissingResourceOptions.with_req(key, base_req))
+    }
+}
 
-impl<P> BufferProvider for LocaleFallbackAdapter<P>
+impl<P> BufferProvider for LocaleFallbackProvider<P>
 where
-    P: LocaleFallbackProvider + BufferProvider,
+    P: BufferProvider,
 {
     fn load_buffer(
         &self,
-        _key: ResourceKey,
-        _req: &DataRequest,
+        key: ResourceKey,
+        base_req: &DataRequest,
```

Comment:
I'd like to, but I already tried that: https://github.com/unicode-org/icu4x/issues/1077. ForkByKeyProvider needs to re-use the request, and that seems like a more performance-sensitive operation than vertical fallback, which is already assumed to carry some cost.

I didn't consider `&mut DataRequest`. That seems risky because it causes a call to `.load_resource` to be non-hermetic; we'd need to put in the contract that providers "reset" the request to its original state.

We could change the signature to return the DataRequest back to the caller upon failure, but then `?` doesn't work as ergonomically. It also carries the same problem as `&mut DataRequest` where we really would like to ensure that the child provider doesn't mutate the request.

Reply:
`DataResponse` already contains the payload and some metadata, it could contain the request as well. In most cases we just to `take_payload` because we don't care about the request (like we currently don't care about the metadata), so this wouldn't change ergonomomics

Reply:
But yeah the mutation case is tricky.

Reply:
This thread is out of scope of this PR since I am not changing the signatures of the traits here.